File: src/eagle/git.js
const { run } = require('./utils')

module.exports = function createGit (opts) {
  const commit = (message) => {
    run('git', ['add', '-A'], opts)
    run('git', ['commit', '-m', message], opts)
  }

  const commitFile = (message, ...files) => {
    run('git', ['commit', '-m', message, '-o', ...files], opts)
  }

  const push = () => {
    run('git', ['push'], opts)
  }

  const pull = () => {
    run('git', ['pull'], opts)
  }

  return Object.freeze({
    commit,
    commitFile,
    push,
    pull
  })
}

File: services/git.go
package services

import (
	"os/exec"
	"sync"
)

type Git struct {
	*sync.Mutex
	Directory string
}

func (g *Git) Commit(msg string) error {
	g.Lock()
	defer g.Unlock()
	cmd := exec.Command("git", "add", "-A")
	cmd.Dir = g.Directory
	err := cmd.Run()

	if err != nil {
		return err
	}

	cmd = exec.Command("git", "commit", "-m", msg)
	cmd.Dir = g.Directory
	err = cmd.Run()
	return err
}

func (g *Git) CommitFile(msg string, files ...string) error {
	g.Lock()
	defer g.Unlock()
	args := []string{"commit", "-m", msg, "--"}
	args = append(args, files...)
	cmd := exec.Command("git", args...)
	cmd.Dir = g.Directory
	err := cmd.Run()
	return err
}

func (g *Git) Push() error {
	g.Lock()
	defer g.Unlock()
	cmd := exec.Command("git", "push")
	cmd.Dir = g.Directory
	err := cmd.Run()
	return err
}

func (g *Git) Pull() error {
	g.Lock()
	defer g.Unlock()
	cmd := exec.Command("git", "pull")
	cmd.Dir = g.Directory
	err := cmd.Run()
	return err
}
